source/grid/grid_out.cc

```diff
@@ -2526,6 +2529,108 @@ void GridOut::write_vtu (const Triangulation<dim,spacedim> &tria,
 
 
 
+template <int dim, int spacedim>
+void GridOut::write_debug_vtu_per_processor (const Triangulation<dim,spacedim> &tria,
+                                             const std::string                 &filename_without_extension,
+                                             const bool                        view_levels,
+                                             const bool                        include_artificial) const
+{
+  std::vector<DataOutBase::Patch<dim,spacedim> > patches;
+  const unsigned int n_datasets=4;
+  std::vector<std::string> data_names;
+  data_names.push_back("level");
+  data_names.push_back("subdomain");
+  data_names.push_back("lvlsubdomain");
+  data_names.push_back("proc_writing");
+
+  const unsigned int n_q_points = GeometryInfo<dim>::vertices_per_cell;
+
+  typename Triangulation<dim, spacedim>::cell_iterator cell, endc;
+  for (cell=tria.begin(), endc=tria.end();
+       cell != endc; ++cell)
+    {
+      if (!include_artificial && cell->level_subdomain_id() ==
+          numbers::artificial_subdomain_id)
+        continue;
+      if (!view_levels && cell->has_children())
+        continue;
+
+      DataOutBase::Patch<dim,spacedim> patch;
+      patch.data.reinit(n_datasets, n_q_points);
+      patch.points_are_available = false;
+
+      for (unsigned int vertex=0; vertex<n_q_points; ++vertex)
+        {
+          patch.vertices[vertex] = cell->vertex(vertex);
+          patch.data(0,vertex) = cell->level();
+          if (!cell->has_children())
+            patch.data(1,vertex) =
+              (double)static_cast<int>(cell->subdomain_id());
+          else
+            patch.data(1,vertex) = -1.0;
+          patch.data(2,vertex) =
+            (double)static_cast<int>(cell->level_subdomain_id());
+          patch.data(3,vertex) = tria.locally_owned_subdomain();
+        }
+
+      for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+        patch.neighbors[f] = numbers::invalid_unsigned_int;
+      patches.push_back(patch);
+    }
+
+  const std::string new_file = (filename_without_extension + ".proc" +
+                                Utilities::int_to_string (tria.locally_owned_subdomain(), 4) +
+                                ".vtu");
+  std::ofstream out(new_file.c_str());
+  std::vector<std_cxx1x::tuple<unsigned int, unsigned int, std::string> > vector_data_ranges;
+  DataOutBase::VtkFlags flags;
+  DataOutBase::write_vtu (patches,
+                          data_names,
+                          vector_data_ranges,
+                          flags,
+                          out);
+  //create .pvtu record
+  if (tria.locally_owned_subdomain() == 0)
+    {
+      std::vector<std::string> filenames;
+
+      //.pvtu needs to reference the files without a relative path because it will be written 
+      //in the same directory. For this, remove any paths from filename.
+      unsigned int indx = filename_without_extension.size();
+      for (int i=indx-1; i>=0; --i)
```

Comment:
there is no reason to have ``i`` and ``indx`` I think. Even better (not tested):
```
size_type pos = filename_without_extension.find_last_of('.');
if (pos == std::npos)
  pos = 0;
else
  pos += 1; // skip '/'

//now use filename_without_extension.substr(pos)
```